specs/commandline/list.md

```diff
@@ -29,6 +29,7 @@ Aliases:
 Flags:
   -d, --debug             debug mode
   -h, --help              help for list
+      --local-artifact    [Preview] list signatures associated with the artifact in OCI layout directory
```

Comment:
1. Can we considering inferring that its a file path? @priteshbandi also brought this up. 
2. Secondly if the flag is needed can we call it something like --file or --path etc. --local is quite subjective.

Reply:
For the option, what if we considered something more flexible than just a boolean. If we use something like --type (maybe this isn't the best name, but I think the concept is interesting) we could specify through the option that it is a directory. That would allow us to easily incorporate new modes for tarballs, SBOMs, or other files that could be added in the future.

---
Example:
```
notation sign $IMAGE
notation sign --type dir /path/to/registry.acme-rockets.io/net-monitor:v2
notation sign --type tar hello-world.tar
...
```

Reply:
@shizhMSFT @patrickzheng200 @qweeah @priteshbandi @vaninrao10 @FeynmanZhou @toddysm @sajayantony please take a look at Kody's comment above. Now we have these options:

1. `--local-content`
2. `--local-artifact`
3. `--path`
4. `--file`
5. `--use-oci-layout`
6. `--oci-dir`
7. `--offline`
8. `file://xxx/`
9. `--dirpath`
10. `--target-path`
11. `--type`

Reply:
In fact, if @yizha1 could remember, the very first version of my design was pretty similar to Kody's suggestion, i.e. specifying type of the user input. So personally speaking, I think it's a good idea as long as there's no usability concern (users need to know what options are available in Notation, and what type their input belong to).

Reply:
I vote for option 3: `--path <dir path | file path>`

Reply:
option 3 +1

Reply:
@yizha1 Question for option 3: we have certificate files, trust policy file and probably local key files in the future, just calling it `--path` might bring in confusion to the user?

Reply:
Looking at the options, can we have a clear understanding what the behavior will be. I think I am confused as a user what will happen if I use one or another option. Here is an example with the `--file` option. If I specify `--file` do I sign the _file_ itself or does Notation considers the _file_ an OCI layout and signs the artifact that is represented by this layout?

The same is the confusion with the `--type tar` - do I sign the actual _tarball_ or does Notation consider the _tar_ an OCI layout and signs the artifact represented by this _tarball_?

Note that there is strong desire to sign individual blobs (and not OCI layouts) and the options above may be quite confusing for users who have that expectation.

Reply:
`--path` and other generic terms are ambiguous if we want to sign a single file in the future. For instance, `--path a.tar` can mean that we are processing the `a.tar` as it is or the content in the `a.tar`.

In fact, I think `--type <type>` makes more sense. `cobra` also allows use to type `--type=<type>` For examples,
- `--type=oci` or `--type=dir` for OCI-Layout in a folder
- `--type=oci` or `--type=tar` for OCI-layout in a tarball
- `--type=file` is reserved for future file signing
- `--type=desc` is reserved for future descriptor signing

However, `--type=<type>` is still not straightforward.

As a reference, [oras](https://github.com/oras-project/oras/blob/v1.0.0/cmd/oras/internal/option/target.go#L65-L72) introduces the following flag:
>     --target type=<type>[[,<key>=<value>][...]]
>  For better UX, the boolean flag `--oci-layout` is introduced as an alias of `--target type=oci-layout`.

Similarly, we could also follow the practice of `oras`.
```sh
notation sign registry.acme-rockets.io/net-monitor:v2                     # remote registry
notation sign --oci-layout registry.acme-rockets.io/net-monitor:v2        # oci-layout folder
notation sign --oci-layout net-monitor.tar:v2 --output net-monitor_v2.sig # oci-layout tarball
notation sign --descriptor desc.json --output desc.sig                    # descriptor
notation sign --file file.bin --output file.sig                           # standalone file

notation list registry.acme-rockets.io/net-monitor:v2              # remote registry
notation list --oci-layout registry.acme-rockets.io/net-monitor:v2 # oci-layout folder
notation list --oci-layout net-monitor.tar:v2                      # oci-layout tarball

notation verify registry.acme-rockets.io/net-monitor:v2                                                           # remote registry
notation verify --scope registry.acme-rockets.io/net-monitor --oci-layout registry.acme-rockets.io/net-monitor:v2 # oci-layout folder
notation verify --scope registry.acme-rockets.io/net-monitor --oci-layout net-monitor.tar:v2                      # oci-layout tarball
notation verify --scope registry.acme-rockets.io/net-monitor --descriptor desc.json --signature desc.sig          # descriptor
notation verify --scope registry.acme-rockets.io/net-monitor --file file.bin --signature file.sig                 # standalone file
```

Reply:
Reading what @shizhMSFT wrote above I think we should look deeper in what we can do with `--type`. Maybe we should rename it to `--artifact-type` and have a predefined list of _types_ that can be used. I would prefer a single switch with multiple pre-defined values instead of multiple switches. 

Looking at what @shizhMSFT wrote above, we currently plan to have the following _types_
- `oci-layout`
- `oci-archive` (does it need to be tar?)
- `descriptor`

In the future, we will have:
- `file` or `blob` although `blob` may be confusing because everything is a `blob`

Reply:
@priteshbandi @vaninrao10  @patrickzheng200 @iamsamirzon @kody-kimberl please take a look at @shizhMSFT and @toddysm's proposal above. Your input is highly appreciated. We'd better finalize it ASAP.

Reply:
I think for usability and compatibility purposes it is better to start with the `--target type=...`. We will need to depecate the `--oci-layout` in the future and customers will also need to change their scripts when we do that. This will make the customer experience more cumbersome.

Reply:
@toddysm the `--oci-layout` will not be deprecated, it is the shorthand for `--target type=oci-layout`. Later when `--target type=<type>` is introduced, users can use either `--target type=oci-layout` or `--oci-layout`. Another reason is, there is not much time left for RC-4, adding `--target type=...` requiring more effort, we will see whether there is a chance between RC-4 and v1. Does it make sense to you? See an example from `docker build buildx -h`
![image](https://user-images.githubusercontent.com/107919912/229096302-82464fda-3674-491e-b633-6b72962423c1.png)

Reply:
@toddysm the `--oci-layout` will not be deprecated, it is the shorthand for `--target type=oci-layout`. Later when `--target type=<type>` is introduced, users can use either `--target type=oci-layout` or `--oci-layout`. Another reason is, there is not much time left for RC-4, adding `--target type=...` requiring more effort, we will see whether there is a chance between RC-4 and v1. Does it make sense to you? See an example from `docker build buildx -h`
![image](https://user-images.githubusercontent.com/107919912/229096302-82464fda-3674-491e-b633-6b72962423c1.png)

Reply:
Agree with adding `--target type=...` flag in v1.0.0. There is only one target type `oci-layout` in RC.4, it's reasonable to use `--oci-layout` as a shorthand for `--target type=oci-layout`.

Reply:
I am not sure having shorthand is a good idea. Having too many switches pollutes the CLI and makes its use hard to use.

Reply:
@toddysm We will consider adding `--target type=` if the time allows for 1.0. For RC.4 it is a good start, since we only have one type now, compared to `--target type=oci-layout`, `--oci-layout` is short and convenient.  Shorthand could be a good practice for some scenarios, if we have multiple choice, and we know what option 80% users will choose, then a shorthand will be convenient for them, since it is short. In our cases, OCI image layout could be the most choice, since it is the only standard now OCI images and other OCI artifacts can be stored in a directory structure under filesystem (could be completely offline), and also easy to copy to remote registry.

Reply:
As discussed, `--oci-layout` and `--scope` will be experimental feature which requires system variable `NOTATION_EXPERIMENTAL` to be set.